File: mindspore/_extends/graph_kernel/expanders/layernorm_grad.py
# Copyright 2020-2021 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ===========================================================================
"""generate json desc for LayerNormGrad"""
from mindspore._extends.graph_kernel.model.model import DataFormat as DF
from ._utils import Expander, ExpanderInfoValidator as VLD


@VLD.add_format(DF.DEFAULT, DF.DEFAULT, DF.DEFAULT, DF.DEFAULT, DF.DEFAULT)
@VLD.check_attrs('begin_norm_axis', 'begin_params_axis')
class LayerNormGrad(Expander):
    """LayerNormGrad expander"""

    def _expand(self, graph_builder):
        x, dy, variance, mean, gamma = self.inputs
        begin_norm_axis = self.attrs['begin_norm_axis']
        begin_params_axis = self.attrs['begin_params_axis']
        epsilon = self.attrs['epsilon'] if 'epsilon' in self.attrs else 1e-11

        if begin_norm_axis < 0:
            begin_norm_axis += len(x.shape)
        if begin_params_axis < 0:
            begin_params_axis += len(x.shape)
        norm_axis = tuple(range(begin_norm_axis, len(x.shape)))
        param_axis = tuple(range(0, begin_params_axis))
        reduce_size = 1.0
        for i in norm_axis:
            reduce_size *= x.shape[i]

        # set some constant val.
        eps = graph_builder.value(x.dtype, epsilon)
        const_one = graph_builder.value(x.dtype, 1.0)
        const_neg_half = graph_builder.value(x.dtype, -0.5)
        const_neg_two = graph_builder.value(x.dtype, -2.0)
        const_two = graph_builder.value(x.dtype, 2.0)
        const_neg_one = graph_builder.value(x.dtype, -1.0)
        mean_cof = graph_builder.value(x.dtype, (1.0 / reduce_size))

        # cal dg db
        var_eps = graph_builder.emit('Add', [variance, eps])
        sqrt_var_eps = graph_builder.emit('Sqrt', [var_eps])
        rsqrt_var_eps = graph_builder.emit('RealDiv', [const_one, sqrt_var_eps])
        x_sub_mean = graph_builder.emit('Sub', [x, mean])
        x_sub_mean_mul_rsqrt_var_eps = graph_builder.emit('Mul', [rsqrt_var_eps, x_sub_mean])
        dg_mul = graph_builder.emit('Mul', [dy, x_sub_mean_mul_rsqrt_var_eps])
        dg = graph_builder.emit('ReduceSum', [dg_mul], attrs={'reduce_axis': param_axis, 'keep_dims': False})
        db = graph_builder.emit('ReduceSum', [dy], attrs={'reduce_axis': param_axis, 'keep_dims': False})

        # cal sum_1
        tmp_var_eps = graph_builder.emit('Mul', [sqrt_var_eps, var_eps])
        r_tmp_var_eps = graph_builder.emit('RealDiv', [const_one, tmp_var_eps])
        x_sub_mean_mul_r_tmp_var_eps = graph_builder.emit('Mul', [x_sub_mean, r_tmp_var_eps])
        dy_mul_gamma = graph_builder.emit('Mul', [dy, gamma])
        tmp_mul = graph_builder.emit('Mul', [dy_mul_gamma, x_sub_mean_mul_r_tmp_var_eps])
        sum_1_mul = graph_builder.emit('Mul', [const_neg_half, tmp_mul])
        sum_1 = graph_builder.emit('ReduceSum', [sum_1_mul], attrs={'reduce_axis': norm_axis, 'keep_dims': True})

        # cal sum_2
        sum_2 = graph_builder.emit('ReduceSum', [dy_mul_gamma], attrs={'reduce_axis': norm_axis, 'keep_dims': True})

        # cal sum_3
        sum_3_mul = graph_builder.emit('Mul', [const_neg_two, x_sub_mean])
        sum_3 = graph_builder.emit('ReduceSum', [sum_3_mul], attrs={'reduce_axis': norm_axis, 'keep_dims': True})

        # cal dx, which is dx1 + dx2 + dx3
        dx_1 = graph_builder.emit('Mul', [dy_mul_gamma, rsqrt_var_eps])
        sum_1_mul_two = graph_builder.emit('Mul', [sum_1, const_two])
        sum_1_mul_two_tmp = graph_builder.emit('Mul', [sum_1_mul_two, mean_cof])
        dx_2 = graph_builder.emit('Mul', [sum_1_mul_two_tmp, x_sub_mean])
        neg_rsqrt_var_eps = graph_builder.emit('Mul', [const_neg_one, rsqrt_var_eps])
        neg_rsqrt_var_eps_mul_sum_2 = graph_builder.emit('Mul', [neg_rsqrt_var_eps, sum_2])
        sum_1_mul_sum_3 = graph_builder.emit('Mul', [sum_1, sum_3])
        mean_cof_mul_sum_1_mul_sum_3 = graph_builder.emit('Mul', [mean_cof, sum_1_mul_sum_3])
        add_tmp = graph_builder.emit('Add', [neg_rsqrt_var_eps_mul_sum_2, mean_cof_mul_sum_1_mul_sum_3])
        dx_3 = graph_builder.emit('Mul', [add_tmp, mean_cof])
        dx_tmp = graph_builder.emit('Add', [dx_1, dx_2])
        dx = graph_builder.emit('Add', [dx_tmp, dx_3])

        return dx, dg, db

File: mindspore/ccsrc/backend/kernel_compiler/cpu/nnacl/fp16_grad/layernorm_grad.c
/**
 * Copyright 2021 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "nnacl/fp16_grad/layernorm_grad.h"
#include <stddef.h>
#include <math.h>

void LayerNormFp16Grad(const float16_t *x, const float16_t *dy, const float16_t *var, const float16_t *mean,
                       const float16_t *gamma, int param_num, int param_size, int block_num, int block_size,
                       float16_t *dx, float16_t *dg, float16_t *db) {
  // var is actually 1/sqrf(var)-> var^0.5
  NNACL_CHECK_ZERO_RETURN(block_size);
  const float16_t *var_sqrt_rev = var;
  for (size_t i = 0; i < param_num; ++i) {
    float dgamma = 0.0f;
    float dbeta = 0.0f;
    for (size_t j = i; j < param_size * param_num; j += param_num) {
      int norm_shift = (int)(j / block_size);
      dgamma += dy[j] * var_sqrt_rev[norm_shift] * (x[j] - mean[norm_shift]);
      dbeta += dy[j];
    }
    dg[i] = (float16_t)dgamma;
    db[i] = (float16_t)dbeta;
  }
  for (size_t i = 0; i < block_num; ++i) {
    float sum1 = 0.0f;
    float sum2 = 0.0f;
    float sum3 = 0.0f;
    for (size_t j = i * block_size; j < (i + 1) * block_size; ++j) {
      int param_shift = j % param_num;
      int norm_shift = (int)(j / block_size);
      float16_t dxm = x[j] - mean[norm_shift];
      float16_t dyg = dy[j] * gamma[param_shift];
      sum1 += -0.5f * dyg * dxm * var_sqrt_rev[norm_shift] * var_sqrt_rev[norm_shift] * var_sqrt_rev[norm_shift];
      sum2 += dyg;
      sum3 += -2.0f * dxm;
    }
    for (size_t j = i * block_size; j < (i + 1) * block_size; ++j) {
      int param_shift = j % param_num;
      int norm_shift = (int)(j / block_size);
      float16_t var_sqrt = var_sqrt_rev[norm_shift];
      float dx1 = dy[j] * gamma[param_shift] * var_sqrt;
      float dx2 = sum1 * 2.0f / block_size * (x[j] - mean[norm_shift]);
      float dx3 = (-1.0f * var_sqrt * sum2 + (1.0f / block_size) * sum1 * sum3) * (1.0f / block_size);
      dx[j] = (float16_t)(dx1 + dx2 + dx3);
    }
  }
}
